flocker/node/agents/blockdevice.py

```diff
@@ -0,0 +1,313 @@
+        for suffix in expected_suffix_list:
+            offset = backing_file.rfind(suffix)
+            if offset > -1:
+                backing_file = backing_file[:offset]
+
+        # Remove the space that may have been between the path and the deleted
+        # flag.
+        backing_file = backing_file.rstrip()
+        backing_file = FilePath(backing_file)
+        devices.append((device_file, backing_file))
+    return devices
+
+
+def losetup_list():
+    """
+    List all the loopback devices on the system.
+
+    :returns: A ``list`` of
+        2-tuple(FilePath(device_file), FilePath(backing_file))
+    """
+    output = check_output(
+        ["losetup", "--all"]
+    ).decode('utf8')
+    return losetup_list_parse(output)
+
+
+def device_for_path(expected_backing_file):
+    """
+    :param FilePath backing_file: A path which may be associated with a
+        loopback device.
+    :returns: A ``FilePath`` to the loopback device if one is found, or
+        ``None`` if no device exists.
+    """
+    for device_file, backing_file in losetup_list():
+        if expected_backing_file == backing_file:
+            return device_file
+    return None
+
+
+@implementer(IBlockDeviceAPI)
+@attributes(['root_path'])
+class LoopbackBlockDeviceAPI(object):
+    """
+    A simulated ``IBlockDeviceAPI`` which creates loopback devices backed by
+    files located beneath the supplied ``root_path``.
+    """
+    _attached_directory_name = 'attached'
+    _unattached_directory_name = 'unattached'
+
+    @classmethod
+    def from_path(cls, root_path):
+        """
+        :param bytes root_path: The path to a directory in which loop back
+            backing files will be created. The directory is created if it does
+            not already exist.
+        :returns: A ``LoopbackBlockDeviceAPI`` with the supplied ``root_path``.
+        """
+        api = cls(root_path=FilePath(root_path))
+        api._initialise_directories()
+        return api
+
+    def _initialise_directories(self):
+        """
+        Create the root and sub-directories in which loopback files will be
+        created.
+        """
+        self._unattached_directory = self.root_path.child(
+            self._unattached_directory_name)
+
+        try:
+            self._unattached_directory.makedirs()
+        except OSError:
+            pass
+
+        self._attached_directory = self.root_path.child(
+            self._attached_directory_name)
+
+        try:
+            self._attached_directory.makedirs()
+        except OSError:
+            pass
+
+    def create_volume(self, size):
+        """
+        Create a file of some size and put it in the ``unattached`` directory.
+
+        See ``IBlockDeviceAPI.create_volume`` for parameter and return type
+        documentation.
+        """
+        volume = BlockDeviceVolume(
+            blockdevice_id=unicode(uuid4()),
+            size=size,
+        )
+        self._unattached_directory.child(
```

Comment:
Eeeeeeeeee. Means we can't test large files without using massive amounts of RAM. I suggest doing this the slightly longer way by opening the file and using... `seek()` and then `write()`? However it is you create giant files.